File: ruby/high-scores/high_scores.rb
# frozen_string_literal: true

# High scores of a single person
class HighScores
  def initialize(scores)
    @scores = scores
  end

  attr_reader :scores

  def latest
    @scores[-1]
  end

  def personal_best
    @scores.max
  end

  def personal_top_three
    @scores.sort.reverse[..2]
  end

  def latest_is_personal_best?
    latest == personal_best
  end
end

File: java/high-scores/src/main/java/HighScores.java
import java.util.Comparator;
import java.util.List;

class HighScores {

  private final List<Integer> highScores;

  public HighScores(List<Integer> highScores) {
    this.highScores = highScores;
  }

  List<Integer> scores() {
    return this.highScores;
  }

  Integer latest() {
    return scores().get(scores().size() - 1);
  }

  Integer personalBest() {
    return scores().stream().max(Comparator.naturalOrder()).orElse(null);
  }

  List<Integer> personalTopThree() {
    return scores().stream().sorted(Comparator.reverseOrder()).limit(3).toList();
  }
}
